app/forms/schools/add_participant_form.rb

```diff
@@ -6,20 +6,114 @@ class AddParticipantForm
     include ActiveModel::Attributes
     include ActiveModel::Serialization
 
+    STEPS = %i[type details choose_mentor email_taken confirm not_implemented].freeze
+
     TYPE_OPTIONS = {
       ect: "Early Career Teacher",
       mentor: "Mentor",
     }.freeze
 
+    attribute :completed_steps, default: []
     attribute :type
+    attribute :full_name
+    attribute :email
+    attribute :school_cohort_id
+    attribute :mentor_id
 
     validates :type,
               on: :type,
               presence: { message: "Please select type of the new participant" },
-              inclusion: { in: TYPE_OPTIONS.keys.map(&:to_s), allow_blank: true }
+              inclusion: { in: TYPE_OPTIONS.keys, allow_blank: true }
+
+    validates :full_name,
+              on: :details,
+              presence: true
+
+    validates :email,
+              on: :details,
+              presence: true,
+              notify_email: { allow_blank: true }
+
+    validates :mentor_id,
+              on: :choose_mentor,
+              presence: true,
+              inclusion: { in: ->(form) { form.mentor_options.map(&:id) + %w[later] } }
 
     def type_options
       TYPE_OPTIONS
     end
+
+    def mentor_options
+      @mentor_options ||= school_cohort.school.mentors
+    end
+
+    def mentor
+      return @mentor if defined? @mentor
+
+      @mentor = (User.find(mentor_id) if mentor_id.present? && mentor_id != "later")
+    end
+
+    def previous_step(current_step)
+      step_index = completed_steps.index(current_step)
+      return completed_steps.last if step_index.nil?
+      return if step_index.zero?
+
+      completed_steps[step_index - 1]
+    end
```

Comment:
You can rejoin the journey halfway through by clicking the "Change" links on the confirmation page - unsure what the desired behaviour of the back button would even be in that case though

Reply:
Yeah, but then we will have the previous steps from the session. You can't enter the journey in the middle without having session with starting step, because if you try you get redirected elsewhere (controller makes sure of it)

Reply:
Oh bay, we will disagree a lot. :P

I am not certain I understand what you mean by leaky. When it return nil then It flags to the controller there is no previous step and it is p to controller to decide what to do in this case. Tthat's the opposite of leaky as it encapusltes the form-related logic without need to know the controller-level logic. If this was to return the back path for the controller to navigate back to, then we can talk about leaked abstraction.

> more generic than it needs to be
Abstract mean generic, so it has to be as generic as possible, with as little assumptions as possible to avoid leaking abstraction. We should not avoid abstraction neither, they are one of the OOD pillars.

Whether we want to start in the middle is only up to the controller and form object should (probably) not really care about this. Even if we decide that form should enforce initial step, than this would only further complcate this PR (happy to do this in refactor as an optional feature). The way this concrete controller is written it is impossible to start the journey midstep for this journey but I might not be the case for some other journeys.

Reply:
> The way this concrete controller is written it is impossible to start the journey midstep for this journey but I might not be the case for some other journeys.

Other journeys using the same form? I guess this goes back to 'ideally, generic multistep form should be in its own PR' so it's easier to understand it without the baggage of feature changes clogging my head :P

Reply:
> Other journeys using the same form?

No, other forms built using the same abstraction. The problem here is that I didn't draw the abstraction bounderies very well -  a lot of decision here were taken based on an implementation of abstract, multistep form and ensuring this abstraction is not leaking, however abstraction itself has not been extracted from the class yet.